
#' Check input and score and scale DE stats.
#'
#' This function takes differential expression gene statistics from scRNAseq
#' data representing a cell type as a dataframe, and assigns scaled scores to
#' the statistic of choice. This is used to rank nodes and edges by differential
#' expression when viewing the bipartite ligand-receptor plots.
#'
#' @param gdb A data frame representing gene statistics from a cell type, where
#'   each row is a gene with official gene symbols as row names. Variables
#'   should be appropriately named statistics or annotations to be included in
#'   the resulting node metadata.
#' @param DEmagn A character vector of length 1 representing the variable name
#'   in the GeneStatList data frames carrying information on the magnitude and
#'   direction of the change of expression for the node (gene) in each cell
#'   type. This is generally a signed logFC or gene expression ratio.
#' @param DEstat A character vector of length 1 representing the variable name
#'   in the GeneStatList data frames carrying information on the statistical
#'   significance of expression change. This is generally a corrected p-value.
#'

CheckDE <- function(gdb,DEmagn,DEstat) {
  if (any( is.na(gdb[[DEmagn]]) )) {
    stop(paste("This function doesn't tolerate missing",
               DEmagn,"values."))
  }
  if (any( is.na(gdb[[DEstat]]) )) {
    stop(paste("This function doesn't tolerate missing",
               DEstat,"values."))
  }
  gdb$DEscaled <- sapply(gdb[[DEmagn]],function(X) {
    if (X == Inf) {
      1.1
    } else if (X == -Inf) {
      -1.1
    } else {
      X / switch(as.character(X >= 0),
                 "TRUE"=max(gdb[[DEmagn]][!is.infinite(gdb[[DEmagn]])]),
                 "FALSE"=min(gdb[[DEmagn]][!is.infinite(gdb[[DEmagn]])]) * -1)
    }
  })
  return(gdb)
}


CheckExpr <- function(gdb,expr) {
  stop("Only DE scoring is supported right now. GeneStatistic argument must be provided.")
}

#' Build cell-cell interaction edge list from scored cell-type statistics
#'
#'
#' @param GeneStatList A named list of dataframes. Each list element should
#'   represent a cell type / cluster to be included in the interaction network,
#'   and should be named accordingly. List elements should contain data frames
#'   where each row is a gene with official gene symbols as row names. Variables
#'   should be appropriately named statistics or annotations to be included in
#'   the resulting node metadata. Variable names should be consistent between
#'   list elements.
#' @param GeneMagnitude A character vector of length 1 representing the variable
#'   name in the GeneStatList data frames carrying information on the magnitude
#'   (and direction of the change) of expression for the node (gene) in each
#'   cell type. This is either a measure of expression (generally mean
#'   expression or detection rate) or a measure of change (signed log expression
#'   ratio a.k.a. logFC).
#' @param GeneStatistic Optional. A character vector of length 1 representing
#'   the variable name in the GeneStatList data frames carrying information on
#'   the statistical significance of expression change. This is generally a
#'   corrected p-value.
#' @param Species Default='hsapiens'. The species of the source data. One of
#'   'hsapiens' or 'mmusculus'. Note that the ligand-receptor database was built
#'   for human, and the mouse version is generated by homology mapping (only
#'   using uniquely mapped homologues).
#'
#' @export

BuildCCInx <- function(GeneStatList,
                       GeneMagnitude,
                       GeneStatistic,
                       Species="hsapiens") {
  if (length(names(GeneStatList)) != length(GeneStatList)) {
    stop("GeneStatList must be a named list where names are cell types.")
  }
  if (any(duplicated(names(GeneStatList)))) {
    stop("GeneStatList names must be unique.")
  }
  if (any(grepl("_",names(GeneStatList)))) {
    stop("GeneStatList names must not contain '_' due to internal naming conventions.")
  }
  message("Checking input data...")
  if (missing(GeneStatistic)) {
    GeneStatList <- pbapply::pbsapply(X=GeneStatList,
                                      FUN=CheckExpr,
                                      expr=GeneMagnitude,
                                      simplify=F)
  } else {
    GeneStatList <- pbapply::pbsapply(X=GeneStatList,
                                      FUN=CheckDE,
                                      DEmagn=GeneMagnitude,
                                      DEstat=GeneStatistic,
                                      simplify=F)
  }

  switch(Species,
         hsapiens=load(system.file("LigRecDB_RData/BaderCCIeditedbyBI.RData",
                                   package="CCInx")),
         mmusculus=load(system.file("LigRecDB_RData/BaderCCIeditedbyBI_mouse.RData",
                                    package="CCInx")),
         stop("Species must be one of 'hsapiens' or 'mmusculus'."))

  tempCN <- c()
  names(GeneStatList) <- gsub("~","_",names(GeneStatList))
  for (a in names(GeneStatList)) {
    for (b in names(GeneStatList)) {
      temp <- paste(sort(c(a,b)),collapse="~")
      if (!temp %in% tempCN) {
        tempCN <- append(tempCN,temp)
      }
    }
  }
  rm(a,b)
  tempComp <- strsplit(tempCN,"~")
  names(tempComp) <- tempCN

  message("Building predicted cell-cell interaction networks...")
  inxL <- pbapply::pbsapply(tempComp,function(Z) {
    a <- Z[1]; b <- Z[2]
    if (nrow(GeneStatList[[a]]) < 1 | nrow(GeneStatList[[b]]) < 1) { return(NULL) }
    inx <- list(edges=c(),
                nodes=c())

    keysAB <- inxDB$key[inxDB$nodeA %in% rownames(GeneStatList[[a]]) &
                          inxDB$nodeB %in% rownames(GeneStatList[[b]])]
    temp_edges <- data.frame(
      sapply(strsplit(keysAB,"_"),function(X)
        paste(paste(X[1],a,sep="_"),paste(X[2],b,sep="_"),sep="~")),
      t(sapply(strsplit(keysAB,"_"),function(X)
        c(paste(X[1],a,sep="_"),paste(X[2],b,sep="_")))),
      stringsAsFactors=F)
    colnames(temp_edges) <- c("key","nodeA","nodeB")
    rownames(temp_edges) <- temp_edges$key

    nodesAB <- unique(c(temp_edges$nodeA,temp_edges$nodeB))
    temp_nodes <- data.frame(nodesAB,
                             do.call(rbind,strsplit(nodesAB,"_")),
                             stringsAsFactors=F)
    colnames(temp_nodes) <- c("node","gene","cellType")
    rownames(temp_nodes) <- temp_nodes$node
    temp_nodes$proteinType <- geneInfo[temp_nodes$gene,"protein_type"]
    if (a == b) {
      temp_nodes <- cbind(temp_nodes,
                          GeneStatList[[a]][temp_nodes[temp_nodes$cellType == a,"gene"],])
    } else {
      temp_nodes <- cbind(temp_nodes,
                          rbind(GeneStatList[[a]][temp_nodes[temp_nodes$cellType == a,"gene"],],
                                GeneStatList[[b]][temp_nodes[temp_nodes$cellType == b,"gene"],]))
    }
    inx$nodes <- rbind(inx$nodes,temp_nodes)

    # temp_edges$meanSigScore <- rowMeans(cbind(inx$nodes[temp_edges$nodeA,"SigScore"],
    #                                           inx$nodes[temp_edges$nodeB,"SigScore"]))
    temp_edges$meanDEscaled <- rowMeans(cbind(inx$nodes[temp_edges$nodeA,"DEscaled"],
                                              inx$nodes[temp_edges$nodeB,"DEscaled"]))
    inx$edges <- rbind(inx$edges,temp_edges)

    keysBA <- inxDB$key[inxDB$nodeA %in% rownames(GeneStatList[[b]]) &
                          inxDB$nodeB %in% rownames(GeneStatList[[a]])]
    temp_edges <- data.frame(
      sapply(strsplit(keysBA,"_"),function(X)
        paste(paste(X[2],a,sep="_"),paste(X[1],b,sep="_"),sep="~")),
      t(sapply(strsplit(keysBA,"_"),function(X)
        c(paste(X[2],a,sep="_"),paste(X[1],b,sep="_")))),
      stringsAsFactors=F)
    colnames(temp_edges) <- c("key","nodeA","nodeB")
    rownames(temp_edges) <- temp_edges$key

    nodesBA <- unique(c(temp_edges$nodeA,temp_edges$nodeB))
    nodesBA <- nodesBA[!nodesBA %in% inx$nodes$node]
    if (length(nodesBA) != 0) {
      temp_nodes <- data.frame(nodesBA,
                               do.call(rbind,strsplit(nodesBA,"_")),
                               stringsAsFactors=F)
      colnames(temp_nodes) <- c("node","gene","cellType")
      rownames(temp_nodes) <- temp_nodes$node
      temp_nodes$proteinType <- geneInfo[temp_nodes$gene,"protein_type"]
      if (a == b) {
        temp_nodes <- cbind(temp_nodes,
                            GeneStatList[[a]][temp_nodes[temp_nodes$cellType == a,"gene"],])
      } else {
        temp_nodes <- cbind(temp_nodes,
                            rbind(GeneStatList[[a]][temp_nodes[temp_nodes$cellType == a,"gene"],],
                                  GeneStatList[[b]][temp_nodes[temp_nodes$cellType == b,"gene"],]))
      }
      inx$nodes <- rbind(inx$nodes,temp_nodes)
    }
    # temp_edges$meanSigScore <- rowMeans(cbind(inx$nodes[temp_edges$nodeA,"SigScore"],
    #                                           inx$nodes[temp_edges$nodeB,"SigScore"]))
    temp_edges$meanDEscaled <- rowMeans(cbind(inx$nodes[temp_edges$nodeA,"DEscaled"],
                                              inx$nodes[temp_edges$nodeB,"DEscaled"]))
    inx$edges <- rbind(inx$edges,temp_edges)

    attr(inx,"GeneMagnitude") <- GeneMagnitude
    attr(inx,"GeneStatistic") <- GeneStatistic

    return(inx)
  },simplify=F)
  return(inxL)
}

